pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java

```diff
@@ -0,0 +1,385 @@
+    throw new UnsupportedOperationException("Consistent-deletion does not support preloading of segments.");
+  }
+
+  @Override
+  protected void doRemoveSegment(IndexSegment segment) {
+    String segmentName = segment.getSegmentName();
+    _logger.info("Removing {} segment: {}, current primary key count: {}",
+        segment instanceof ImmutableSegment ? "immutable" : "mutable", segmentName, getNumPrimaryKeys());
+    long startTimeMs = System.currentTimeMillis();
+
+    try (
+        UpsertUtils.PrimaryKeyReader primaryKeyReader = new UpsertUtils.PrimaryKeyReader(segment, _primaryKeyColumns)) {
+      removeSegment(segment,
+          UpsertUtils.getPrimaryKeyIterator(primaryKeyReader, segment.getSegmentMetadata().getTotalDocs()));
+    } catch (Exception e) {
+      throw new RuntimeException(
+          String.format("Caught exception while removing segment: %s, table: %s", segment.getSegmentName(),
+              _tableNameWithType), e);
+    }
+
+    // Update metrics
+    long numPrimaryKeys = getNumPrimaryKeys();
+    updatePrimaryKeyGauge(numPrimaryKeys);
+    _logger.info("Finished removing segment: {} in {}ms, current primary key count: {}", segmentName,
+        System.currentTimeMillis() - startTimeMs, numPrimaryKeys);
+  }
+
+  @Override
+  protected void removeSegment(IndexSegment segment, Iterator<PrimaryKey> primaryKeyIterator) {
+    Set<Object> deDuplicatedPrimaryKeys = new HashSet<>();
+    while (primaryKeyIterator.hasNext()) {
+      PrimaryKey primaryKey = primaryKeyIterator.next();
+      _primaryKeyToRecordLocationMap.computeIfPresent(HashUtils.hashPrimaryKey(primaryKey, _hashFunction),
+          (pk, recordLocation) -> {
+            if (deDuplicatedPrimaryKeys.contains(pk)) {
+              return recordLocation;
+            }
+            deDuplicatedPrimaryKeys.add(pk);
+            if (recordLocation.getSegment() == segment) {
+              return null;
+            }
+            return new RecordLocation(recordLocation.getSegment(), recordLocation.getDocId(),
+                recordLocation.getComparisonValue(),
+                RecordLocation.decrementSegmentCount(recordLocation.getDistinctSegmentCount()));
+          });
+    }
+  }
+
+  @Override
+  public void doRemoveExpiredPrimaryKeys() {
+    AtomicInteger numDeletedTTLKeysRemoved = new AtomicInteger();
+    double largestSeenComparisonValue = _largestSeenComparisonValue.get();
+    double deletedKeysThreshold;
+    if (_deletedKeysTTL > 0) {
+      deletedKeysThreshold = largestSeenComparisonValue - _deletedKeysTTL;
+    } else {
+      deletedKeysThreshold = Double.MIN_VALUE;
+    }
+
+    _primaryKeyToRecordLocationMap.forEach((primaryKey, recordLocation) -> {
+      double comparisonValue = ((Number) recordLocation.getComparisonValue()).doubleValue();
+      if (_deletedKeysTTL > 0 && comparisonValue < deletedKeysThreshold
+          && recordLocation.getDistinctSegmentCount() <= 1) {
```

Comment:
perhaps add some comments on this check, or add comment on the disctinctSegmentCount field in  recordLocation that it's used here, as it's kinda critical for the correctness of this feature.

also looks like a lot of methods are same with ConcurrentMapPartitionUpsertMetadataManager. How about extracting a base class like BaseConcurrentMapPartitionUpsertMetadataManager, for both to reuse.